File: Templates/Project Templates/macOS/Application Extension/Photo Project Extension.xctemplate/PhotoProjectViewController.m
//
//  ___FILENAME___
//  ___PACKAGENAME___
//
//  Created by ___FULLUSERNAME___ on ___DATE___.
//  ___COPYRIGHT___
//

#import "___FILEBASENAME___.h"
#import "ProjectTypeDescriptionDataSource.h"
#import <PhotosUI/PhotosUI.h>

@interface ___FILEBASENAME___ () <PHProjectExtensionController>
@end

@implementation ___FILEBASENAME___

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
}

#pragma mark - PHProjectExtensionController

- (id<PHProjectTypeDescriptionDataSource>)typeDescriptionDataSourceForCategory:(PHProjectCategory)category invalidator:(id<PHProjectTypeDescriptionInvalidator>)invalidator {
    return [[ProjectTypeDescriptionDataSource alloc] init];
}

- (void)beginProjectWithExtensionContext:(PHProjectExtensionContext *)extensionContext projectInfo:(PHProjectInfo *)projectInfo completion:(nonnull void (^)(NSError * ))completion {
    dispatch_async(dispatch_get_main_queue(), ^(void) {
        // do initialization here
        completion(nil);
    });
}

- (void)resumeProjectWithExtensionContext:(PHProjectExtensionContext *)extensionContext completion:(nonnull void (^)(NSError * ))completion {
    dispatch_async(dispatch_get_main_queue(), ^(void) {
        // do initialization here
        completion(nil);
    });
}

- (void)finishProjectWithCompletionHandler:(void (^)(void))completion {
    // do any finalization here
    completion();
}

#pragma mark - Helper

- (PHProjectExtensionContext *)projectExtensionContext {
    return (PHProjectExtensionContext *)self.extensionContext;
}


@end

File: Templates/Project Templates/macOS/Application Extension/Photo Project Extension.xctemplate/PhotoProjectViewController.swift
//
//  ___FILENAME___
//  ___PACKAGENAME___
//
//  Created by ___FULLUSERNAME___ on ___DATE___.
//  ___COPYRIGHT___
//

import Cocoa
import PhotosUI

class PhotoProjectViewController: NSViewController, PHProjectExtensionController {
    
    var projectExtensionContext: PHProjectExtensionContext? {
        get {
            return extensionContext as? PHProjectExtensionContext
        }
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do view setup here.
    }
    
    // MARK: - PHProjectExtensionController
    
    func typeDescriptionDataSource(for category: PHProjectCategory, invalidator: PHProjectTypeDescriptionInvalidator) -> PHProjectTypeDescriptionDataSource {
        return ProjectTypeDescriptionDataSource()
    }
    
    func beginProject(with extensionContext: PHProjectExtensionContext, projectInfo: PHProjectInfo, completion: @escaping (Error?) -> Void) {
        DispatchQueue.main.async {
            // do initialization here
            completion(nil)
        }
    }
    
    func resumeProject(with extensionContext: PHProjectExtensionContext, completion: @escaping (Error?) -> Void) {
        DispatchQueue.main.async {
            // do initialization here
            completion(nil)
        }
    }
    
    func finishProject(completionHandler completion: @escaping () -> Void) {
        // do any finalization here
        completion()
    }
}
